src/uct/cuda/cuda_copy/cuda_copy_ep.c

```diff
@@ -108,45 +144,41 @@ uct_cuda_copy_post_cuda_async_copy(uct_ep_h tl_ep, void *dst, void *src,
     ucs_status_t status;
     ucs_memory_type_t src_type;
     ucs_memory_type_t dst_type;
-    CUstream *stream;
+    CUstream stream;
     ucs_queue_head_t *event_q;
+    uct_cuda_copy_ctx_rsc_t *ctx_rsc;
 
     if (!length) {
         return UCS_OK;
     }
 
-    /* ensure context is set before creating events/streams */
-    if (iface->cuda_context == NULL) {
-        UCT_CUDADRV_FUNC_LOG_ERR(cuCtxGetCurrent(&iface->cuda_context));
-        if (iface->cuda_context == NULL) {
-            ucs_error("attempt to perform cuda memcpy without active context");
-            return UCS_ERR_IO_ERROR;
-        }
+    status = uct_cuda_copy_get_ctx_rsc(iface, &ctx_rsc);
+    if (status != UCS_OK) {
+        return status;
     }
 
     src_type = uct_cuda_copy_get_mem_type(base_iface->md, src, length);
     dst_type = uct_cuda_copy_get_mem_type(base_iface->md, dst, length);
-    q_desc   = &iface->queue_desc[src_type][dst_type];
+    q_desc   = &ctx_rsc->queue_desc[src_type][dst_type];
     event_q  = &q_desc->event_queue;
-    stream   = uct_cuda_copy_get_stream(iface, src_type, dst_type);
+    stream   = uct_cuda_copy_get_stream(ctx_rsc, src_type, dst_type);
     if (stream == NULL) {
         ucs_error("stream for src %s dst %s not available",
                    ucs_memory_type_names[src_type],
                    ucs_memory_type_names[dst_type]);
         return UCS_ERR_IO_ERROR;
     }
 
-    cuda_event = ucs_mpool_get(&iface->cuda_event_desc);
+    cuda_event = ucs_mpool_get(&ctx_rsc->cuda_event_desc);
     if (ucs_unlikely(cuda_event == NULL)) {
         ucs_error("Failed to allocate cuda event object");
         return UCS_ERR_NO_MEMORY;
     }
 
     status = UCT_CUDADRV_FUNC_LOG_ERR(
-            cuMemcpyAsync((CUdeviceptr)dst, (CUdeviceptr)src, length, *stream));
+            cuMemcpyAsync((CUdeviceptr)dst, (CUdeviceptr)src, length, stream));
 
-    status = UCT_CUDADRV_FUNC_LOG_ERR(
-            cuEventRecord(cuda_event->event, *stream));
+    status = UCT_CUDADRV_FUNC_LOG_ERR(cuEventRecord(cuda_event->event, stream));
```

Comment:
is it possible that `cuMemcpyAsync` fails, but `cuEventRecord` succeeds? 
`status` will b eoverwritten then